index.js

```diff
@@ -61,7 +61,16 @@ function got(url, opts, cb) {
 	}
 
 	if (body) {
+		if (typeof body !== 'string' && !Buffer.isBuffer(body) && !isStream.readable(body)) {
+			throw new GotError('options.body must be a ReadableStream, string or Buffer');
+		}
+
 		opts.method = opts.method || 'POST';
+
+		if (!opts.headers['transfer-encoding'] && !isStream.readable(body)) {
+			var length = typeof body === 'string' ? Buffer.byteLength(body) : body.length;
+			opts.headers['content-length'] = opts.headers['content-length'] || length;
```

Comment:
You could test the presence of `content-length` directly in the `if` above instead.